File: algorithm/HDU-ACM/ambition0109/1297.java
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: ambition0109
////Nickname: Ambition
////Run ID: 
////Submit time: 2010-10-30 15:56:41
////Compiler: Java
//////////////////////////////////////////////////////
////Problem ID: 1297
////Problem Title: 
////Run result: Accept
////Run time:171MS
////Run memory:7652KB
//////////////////System Comment End//////////////////
import java.util.*;
import java.math.*;

public class Main {
	public static void main(String[] args){
		BigInteger[] num=new BigInteger[1005];
		num[0]=BigInteger.ONE;
		num[1]=BigInteger.ONE;
		num[2]=BigInteger.valueOf(2);
		num[3]=BigInteger.valueOf(4);
		for(int i=4;i<=1000;i++){
			num[i]=num[i-4];
			num[i]=num[i].add(num[i-2]);
			num[i]=num[i].add(num[i-1]);
		}
		Scanner in=new Scanner(System.in);
		int n;
		while(in.hasNextInt()){
			n=in.nextInt();
			System.out.println(num[n]);
		}
	}
}

File: algorithm/HDU-ACM/ambition0109/1297.cpp
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: ambition0109
////Nickname: Ambition
////Run ID: 
////Submit time: 2010-04-08 16:17:25
////Compiler: Visual C
//////////////////////////////////////////////////////
////Problem ID: 1297
////Problem Title: 
////Run result: Accept
////Run time:15MS
////Run memory:4124KB
//////////////////System Comment End//////////////////
#include<stdio.h>
int f[1001][1000];
int main()
{
    int n,i,k=1,j;
    for(i=0;i<1001;i++)
        for(j=0;j<1000;j++)
            f[i][j]=0;
    f[0][0]=f[1][0]=f[2][0]=f[3][0]=1;
    f[0][1]=1;f[1][1]=1;f[2][1]=2;f[3][1]=4;
    for(i=4;i<1001;i++)
    {
        for(j=1;j<=k;j++)
        {
            f[i][j]+=f[i-1][j]+f[i-2][j]+f[i-4][j];
            if(f[i][j]>9)
            {
                f[i][j+1]+=(f[i][j]/10);
				f[i][j]%=10;
            }
        }
        if(f[i][k+1]!=0) k++;
        f[i][0]=k;
    }
    while(scanf("%d",&n)!=EOF)
    {
        for(i=f[n][0];i>0;i--)
            printf("%d",f[n][i]);
        printf("\n");
    }
    return 0;
}